### PR TITLE
[geeliette] iP

### PR DESCRIPTION
**Pull request by geeliette**
# Fleur: Your Key To Blooming 🌸 
> "A good system shortens the road to the goal." - Orison Swett Marden

**Fleur** is the system for you to keep track of all your daily tasks, so that you can achieve your goals more efficiently. 

## Why You'll Love Fleur
- <ins>Text-based</ins> 📄 : Simple and straightforward interaction
- <ins>User-friendly</ins> 🩵 : Intuitive commands that anyone can learn
- <ins>Lightning fast</ins> ⚡ : Instant feedback and speedy task management

## How To Use It
1. Download Fleur from [here](https://github.com/geeliette/ip/releases/download/A-Release/fleur.jar)
2. Double-click it or run it using the command: `java -jar "fleur.jar"`
3. Start adding your tasks
4. Sit back and enjoy a more organised life! 🦔 

And it's ***FREE***!

## Features
- [x] Manage tasks (ToDos, Deadlines and Events)
- [x] Find tasks using keywords
- [ ] Reminders (coming soon)

## Main Method
If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        new Fleur("fleur.txt").run();
    }
}

### CODE REVIEW

File: src/main/java/fleur/Parser.java [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+            } else if (str.startsWith("delete")) {
+                int index = Integer.parseInt(str.substring(7)) - 1;
+                Task deletedTask = this.taskList.getTask(index);
+                System.out.println("D'accord, I 'ave removed zis task from your list:");
+                System.out.println(deletedTask.toString());
+                this.taskList.deleteTask(index);
+                System.out.println("Now you 'ave " + this.taskList.size() + " task(s) in your list.");
+            } else {
+                throw new FleurInvalidCommandException();
+            }
```
Comment (charlenelau-20) [inline src/main/java/fleur/Parser.java]: Maybe using switch case statements could make the code easier to read?

File: src/main/java/fleur/Storage.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+                    break;
+                case 'E':
+                    String[] arr = str.substring(7).split("\\(from: ");
+                    String description = arr[0];
+                    String from = arr[1].split("to: ")[0].trim();
+                    String to = arr[1].split("to: ")[1].replace(")", "");
+                    LocalDate dateFrom = LocalDate.parse(from, INPUT);
+                    LocalDate dateTo = LocalDate.parse(to, INPUT);
+                    t = new Event(description, dateFrom, dateTo);
+                    break;
```
Comment (charlenelau-20) [inline src/main/java/fleur/Storage.java]: Just a minor layout fix; remember not to have indentations for case clauses in this course!

File: src/main/java/fleur/Ui.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+public class Ui {
+
+    private static Scanner scanner;
+
+    public Ui() {
+        this.scanner = new Scanner(System.in);
+    }
+
+    public void welcomeMessage() {
```
Comment (charlenelau-20) [inline src/main/java/fleur/Ui.java]: Perhaps using a verb for the method name, e.g. showWelcomeMessage(), could make it more intuitive.

File: src/main/java/fleur/Storage.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+                    t = new ToDo(str.substring(7));
+                    break;
+                case 'D':
+                    String desc = str.substring(7).split("\\(by: ")[0];
+                    String date = str.substring(7).split("\\(by: ")[1].replace(")", "");
+                    LocalDate by = LocalDate.parse(date, INPUT);
+                    t = new Deadline(desc, by);
+                    break;
+                case 'E':
+                    String[] arr = str.substring(7).split("\\(from: ");
```
Comment (charlenelau-20) [inline src/main/java/fleur/Storage.java]: Maybe using full/more descriptive words for the variable names 'desc' and 'arr' would make it easier for readers to understand what they are used for.

**Review by charlenelau-20 — COMMENTED** [review]
Looks good overall! Also liked how unique the responses are!

File: src/main/java/fleur/Storage.java [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+public class Storage {
+
+    private static String dataFile;
+
+    public Storage(String dataFile) {
+        this.dataFile = dataFile;
+    }
+
+    protected ArrayList<Task> loadData() throws IOException, FleurCorruptFileException {
```
Comment (kai824) [inline src/main/java/fleur/Storage.java]: > `protected`

Why do you choose to use protected here? Will there be any class that will inherit from Storage and need access to this method?

File: src/main/java/fleur/Fleur.java [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+//    private void exit() {
+//        try {
+//            saveData();
+//            System.out.println("Au revoir, 'ope to see you again soon!");
+//        } catch (IOException e) {
+//            System.out.println("Mince! Zere was an error when saving your tasks.");
+//        }
+//    }
+//
+//}
```
Comment (kai824) [inline src/main/java/fleur/Fleur.java]: Having such a block of commented code may not be the best practice. Consider deleting it

File: src/main/java/fleur/Parser.java [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+            String[] arr = str.substring(6).split("/from");
+            String desc = arr[0];
+            String from = arr[1].split("/to")[0].trim();
+            String to = arr[1].split("/to")[1].trim();
+            LocalDate dateFrom = LocalDate.parse(from, INPUT);
+            LocalDate dateTo = LocalDate.parse(to, INPUT);
+            Task t = new Event(desc, dateFrom, dateTo);
+            this.taskList.addTask(t);
+            System.out.println("Bah, oui! I 'ave added zis event task to your list:");
+            System.out.println(t.toString());
```
Comment (kai824) [inline src/main/java/fleur/Parser.java]: While this Task object is temporary, consider using a more descriptive name (e.g. `taskToAdd`)

File: src/main/java/fleur/Parser.java [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+            System.out.println("Now you 'ave " + this.taskList.size() + " task(s) in your list.");
+        } catch (IndexOutOfBoundsException e) {
+            throw new FleurMissingDetailsException();
+        } catch (DateTimeParseException e) {
+            throw new FleurInvalidDateException();
+        }
+    }
+
+    private void addEvent(String str) throws FleurMissingDetailsException, FleurInvalidDateException {
```
Comment (kai824) [inline src/main/java/fleur/Parser.java]: Good job specifying the two checked Exceptions separately instead of just specifying `throws FleurException`, so that the callers can choose to handle them separately.

File: src/main/java/fleur/Storage.java (hunk 2 above)
Comment (kai824) [inline src/main/java/fleur/Storage.java]: Remember to configure your IDE not to have indentations for case clauses for this course

File: text-ui-test/input.txt [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+todo tutorial
+list
+mark 2
+mark 1
+list
+unmark 2
+list
+delete 3
+hi
+todo
```
Comment (kai824) [inline text-ui-test/input.txt:12]: Good job testing the edge cases and error messages! Consider testing the case of index out of bounds as well

File: src/main/java/fleur/FleurException.java [hunk 10]
```diff
@@ -0,0 +1,7 @@
+package fleur;
+
+public class FleurException extends Exception {
```
Comment (kai824) [inline src/main/java/fleur/FleurException.java]: Perhaps you could group the Exception classes together into 1 package to make the file organisation even tidier

**Review by kai824 — COMMENTED** [review]
Looks good overall! Love the creativity in the use of French in your responses.

Do remember to add javadoc comments.
